### PR TITLE
Add nest, unnest, improve flatten

### PR DESCRIPTION
**Pull request by bkamins**
Fixes
https://github.com/JuliaData/DataFrames.jl/issues/3005
https://github.com/JuliaData/DataFrames.jl/issues/2890
https://github.com/JuliaData/DataFrames.jl/issues/3116
https://github.com/JuliaData/DataFrames.jl/issues/2767
https://github.com/JuliaData/DataFrames.jl/issues/2767

The PR adds `nest` and `unnest` and introduces `scalar` kwarg to `flatten` (which is needed in `unnest`.

`flatten` is ready for review.

For `nest` and `unnest` requires discussion if we like the proposed API (they work, but maybe we will decide to change API).

Some important decisions I propose:
* `nest` only works on `GroupedDataFrame` (the reason is to avoid complexity of group order specification); nesting is done always to `DataFrame` (to keep things simple); another not easy decision is syntax I proposed `[:x, :y] => :z` which means that columns `:x` and `:y` should be nested and stored in column `:z` (but I would like to confirm that we find it intuitive, as syntax `:z => [:x, :y]` also could be advocated for).
* `unnest` supports both tables (e.g. `DataFrame`) and rows (e.g. `Tables.AbstractRow`) and has two options: `flatten=true`, when rows of the nested columns are flattened, and `flatten=false` (when they are left as is - this is probably useful, if we work with rows)

TODO:

- add metadata
- write tests
- update manual

CC @nalimilan @pdeffebach @jariji

### CODE REVIEW

File: src/abstractdataframe/nest.jl [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+`cols` (default `:setequal`) and `promote` (default `true`) keyword arguments
+have the same meaning as in [`push!`](@ref).
+
+If `makeunique=false` (the default) produced column names must be unique.
+If `makeunique=true` then duplicate column names will be suffixed with `_i`
+(`i` starting at `1` for the first duplicate).
+
+If `flatten=true` (the default) then newly created columns are flattened
+using [`flatten`](@ref) with `scalar=Missing` keyword argument.
+
```
Comment (jariji) [inline src/abstractdataframe/nest.jl:176]: Missing examples

Comment (bkamins) [reply src/abstractdataframe/nest.jl:176]: yes - I first want to discuss the API.

File: src/abstractdataframe/nest.jl [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+julia> df3 = DataFrame(a=[1, 2], b=[[1, 2], [3, 4]], c=[[5, 6], [7]])
+2×3 DataFrame
+ Row │ a      b       c
+     │ Int64  Array…  Array…
+─────┼───────────────────────
+   1 │     1  [1, 2]  [5, 6]
+   2 │     2  [3, 4]  [7]
+
+julia> flatten(df3, [:b, :c])
```
Comment (jariji) [inline src/abstractdataframe/nest.jl:289]: Is `equal-length-zip` the desired behavior here? The alternatives that come to mind are `zip-shortest`, `zip-longest`, and especially `cartesian-product`.

File: src/abstractdataframe/nest.jl (hunk 1 above)
Comment (bkamins) [reply src/abstractdataframe/nest.jl:176]: If we were to discuss the case that I think might be most problematic is:
```
julia> df = DataFrame(x=[(a="aa", b="bb"), (a="pp", b="qq")])
2×1 DataFrame
 Row │ x
     │ NamedTup…
─────┼──────────────────────
   1 │ (a = "aa", b = "bb")
   2 │ (a = "pp", b = "qq")

julia> unnest(df, :x, flatten=false)
2×2 DataFrame
 Row │ a       b      
     │ String  String 
─────┼────────────────
   1 │ aa      bb
   2 │ pp      qq

julia> unnest(df, :x)
4×2 DataFrame
 Row │ a     b    
     │ Char  Char 
─────┼────────────
   1 │ a     b
   2 │ a     b
   3 │ p     q
   4 │ p     q
```
so maybe we should allow for passing `scalar` kwarg here also or not allow to flatten at all (but expect it to be a separate step).

File: src/abstractdataframe/nest.jl (hunk 2 above)
Comment (bkamins) [reply src/abstractdataframe/nest.jl:289]: This is what is the current behavior. We could add a kwarg for other behaviors (but I would discuss them in a separate issue, for creating cartesian product we have `allcombinations`).

File: src/abstractdataframe/nest.jl (hunk 1 above)
Comment (jariji) [reply src/abstractdataframe/nest.jl:176]: What would be the output of that operation with `scalar`?

Comment (jariji) [reply src/abstractdataframe/nest.jl:176]: I'm not clear on why `unnest` supports unpacking single-row cells. I expected it to work on Table-valued cells, not row-valued ones. That seems like a different operation.

Comment (bkamins) [reply src/abstractdataframe/nest.jl:176]: > What would be the output of that operation with scalar?

If we allowed passing `scalar` user could write:

```
julia> unnest(df, :x, scalar=String)
2×2 DataFrame
 Row │ a       b      
     │ String  String 
─────┼────────────────
   1 │ aa      bb
   2 │ pp      qq
```
and it would not get flattened.

---

> I expected it to work on Table-valued cells, not row-valued ones. That seems like a different operation.

I assume you have experience with dplyr, where there is no distinction between table-values and row-values. In JuliaData ecosystem there is such a distinction, and both operations are useful, and are essentially `unnest` (and we need to make a design that fits JuliaData ecosystem, not just mimics dplyr). Here is an example of the difference:

```
julia> df = DataFrame(data=[(a=1, b=2), (a=3, c=4)])
2×1 DataFrame
 Row │ data
     │ NamedTup…      
─────┼────────────────
   1 │ (a = 1, b = 2)
   2 │ (a = 3, c = 4)

julia> unnest(df, :data, flatten=false, cols=:union)
2×3 DataFrame
 Row │ a      b        c       
     │ Int64  Int64?   Int64?  
─────┼─────────────────────────
   1 │     1        2  missing 
   2 │     3  missing        4

julia> df = DataFrame(data=[(a=1:2, b=3:4), (a=5:6, c=7:8)])
2×1 DataFrame
 Row │ data
     │ NamedTup…
─────┼────────────────────
   1 │ (a = 1:2, b = 3:4)
   2 │ (a = 5:6, c = 7:8)

julia> unnest(df, :data, flatten=true, cols=:union)
4×3 DataFrame
 Row │ a      b        c       
     │ Int64  Int64?   Int64?  
─────┼─────────────────────────
   1 │     1        3  missing 
   2 │     2        4  missing 
   3 │     5  missing        7
   4 │     6  missing        8

julia> unnest(df, :data, flatten=false, cols=:union)
2×3 DataFrame
 Row │ a          b           c
     │ UnitRang…  UnitRang…?  UnitRang…?
─────┼───────────────────────────────────
   1 │ 1:2        3:4         missing
   2 │ 5:6        missing     7:8
```
As you can see for row-like input only one operation makes sense (`flatten=false`), for table-like input two operations make sense (`flatten=true`, and `flatten=false`). All three options have been asked for by the users, so I propose design that supports all of them. We could indeed define two separate functions (one for row-data and one for table-data), but it would be almost identical (with the difference that for row data there would be no `flatten` kwarg, but for table-data still `flatten` would be needed).

My initial thought if we need to provide `flatten` keyword argument at all as it is the same as calling `flatten` in the last step. But maybe we should for convenience support it, and then maybe it is useful to allow passing `scalar` keyword argument.

Comment (jariji) [reply src/abstractdataframe/nest.jl:176]: I don't like `flatten=`. In general, rather than `f_and_maybe_g(;g=true)` I prefer `g(f())`. Especially here where so much meaning depends on that kwarg.

---

In some other issue, a function `union` was preferred over the symbol `:union`. Perhaps here, too?

Comment (bkamins) [reply src/abstractdataframe/nest.jl:176]: > I don't like `flatten=`.

I also have this feeling. Maybe we could define `unnest` that works as `flatten=true` and e.g. `spread` that works as `flatten=false` now.

In practice `unnest` then would call `spread` and then `flatten` (the convenience of this is that it would automatically know which columns should be flattened).

---
Regarding `cols` kwarg it supports the following operations:
```
    •  If cols == :setequal (this is the default) then row must contain exactly the same columns as df (but possibly in a different order).

    •  If cols == :orderequal then row must contain the same columns in the same order (for AbstractDict this option requires that keys(row) matches
       propertynames(df) to allow for support of ordered dicts; however, if row is a Dict an error is thrown as it is an unordered collection).

    •  If cols == :intersect then row may contain more columns than df, but all column names that are present in df must be present in row and only
       they are used to populate a new row in df.

    •  If cols == :subset then the behavior is like for :intersect but if some column is missing in row then a missing value is pushed to df.

    •  If cols == :union then columns missing in df that are present in row are added to df (using missing for existing rows) and a missing value is
       pushed to columns missing in row that are present in df.
```
I would keep them as is, as changing this would be breaking (+ some of them are not easily expressed as functions).

File: src/abstractdataframe/nest.jl [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+to `col` becomes a longer vector where the original entries are concatenated.
+Elements of row `i` of `df` in columns other than `cols` will be repeated
+according to the length of `df[i, col]`. These lengths must therefore be the
+same for each `col` in `cols`, or else an error is raised. Note that these
+elements are not copied, and thus if they are mutable changing them in the
+returned `DataFrame` will affect `df`.
+
+`cols` can be any column selector ($COLUMNINDEX_STR; $MULTICOLUMNINDEX_STR).
+
+If `scalar` is passed then values that have this type in flattened columns
```
Comment (jariji) [inline src/abstractdataframe/nest.jl:225]: Will I need to explicitly pass `scalar=Union{String,Missing}` for such columns to avoid splatting strings into chars?

Comment (bkamins) [reply src/abstractdataframe/nest.jl:225]: or e.g. `scalar=AbstractString` if you do not care about `missing`.

We could make `scalar=Union{Missing, AbstractString}` the default, but it would be breaking (as these are the most common cases people would care about). Still - maybe we could allow for it as it is only slightly breaking and it seems most people feel this behavior is preferred.

As a starting point I proposed a non-breaking design (i.e. doing what the function does now by default, but allowing a different behavior if needed).

Comment (jariji) [reply src/abstractdataframe/nest.jl:225]: I think changing the default is better. I'm generally annoyed that strings are iterable and I'd rather not have to remember that they are.

Comment (bkamins) [reply src/abstractdataframe/nest.jl:225]: I agree. But the tension here is "better" vs "breaking". Therefore this needs a consensus in JuliaData community.

In general we have three options for `AbstractString` as default:
* expand (the current behavior)
* error (special case `AbstractString` in the code and throw an error if someone tries to expand it, unless it is included in `scalar`)
* treat as scalar by default (and broadcast if needed)

File: src/abstractdataframe/nest.jl [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+
+`cols` (default `:setequal`) and `promote` (default `true`) keyword arguments
+have the same meaning as in [`push!`](@ref).
+
+If `makeunique=false` (the default) produced column names must be unique.
+If `makeunique=true` then duplicate column names will be suffixed with `_i`
+(`i` starting at `1` for the first duplicate).
+
+If `flatten=true` (the default) then newly created columns are flattened
+using [`flatten`](@ref) with `scalar=Missing` keyword argument.
```
Comment (nalimilan) [inline src/abstractdataframe/nest.jl:175]: Why do we need to special case `Missing`?

File: src/abstractdataframe/nest.jl [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+                makeunique::Bool=false, flatten::Bool=true)
+    ref_df = select(df, Not(collect(Any, src)))
+    col_count = ncol(ref_df)
+    for idx in src
+        col = df[!, idx]
+        tmp_df = DataFrame()
+        for v in col
+            if v isa DataFrame # produce DataFrameRow
+                # if flatten=false make a copy to avoid aliases
+                v = DataFrame([n => [flatten ? c : copy(c)]
```
Comment (nalimilan) [inline src/abstractdataframe/nest.jl:193]: When `flatten=true`, isn't this line unnecessary?

Also I don't get what `only` does here.

File: src/abstractdataframe/nest.jl [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+                            AbstractVector{<:AbstractString}}=:setequal,
+                promote::Bool=(cols in [:union, :subset]),
+                makeunique::Bool=false, flatten::Bool=true)
+    ref_df = select(df, Not(collect(Any, src)))
+    col_count = ncol(ref_df)
+    for idx in src
+        col = df[!, idx]
+        tmp_df = DataFrame()
+        for v in col
+            if v isa DataFrame # produce DataFrameRow
```
Comment (nalimilan) [inline src/abstractdataframe/nest.jl:191]: I don't see where `DataFrameRow` appears here.

**Comment by nalimilan** [comment]
> `nest` only works on `GroupedDataFrame` (the reason is to avoid complexity of group order specification); nesting is done always to `DataFrame` (to keep things simple); another not easy decision is syntax I proposed `[:x, :y] => :z` which means that columns `:x` and `:y` should be nested and stored in column `:z` (but I would like to confirm that we find it intuitive, as syntax `:z => [:x, :y]` also could be advocated for).

Agreed. `[:x, :y] => :z` sounds more logical as the input is on the LHS and the output on the RHS.

> `unnest` supports both tables (e.g. `DataFrame`) and rows (e.g. `Tables.AbstractRow`) and has two options: `flatten=true`, when rows of the nested columns are flattened, and `flatten=false` (when they are left as is - this is probably useful, if we work with rows)

Let's continue discussion at https://github.com/JuliaData/DataFrames.jl/pull/3258#discussion_r1059102805.

File: src/abstractdataframe/nest.jl (hunk 1 above)
Comment (bkamins) [reply src/abstractdataframe/nest.jl:176]: `flatten=false` is in my opinion, in the context of DataFrames.jl more natural. I know that dplyr users expect `flatten=true`, but a natural use-case for data-frames is:
1. you get a source JSON file
2. your read it in into a data frame and get something like:
```
julia> df = DataFrame(col=[(a="x", b=1), (a="y", c=[1,2])])
2×1 DataFrame
 Row │ col
     │ NamedTup…
─────┼───────────────────────
   1 │ (a = "x", b = 1)
   2 │ (a = "y", c = [1, 2])
```
3. you want that tuned into:
```
julia> df = DataFrame(a=["x", "y"], b=[1, missing], c=[missing, [1,2]])
2×3 DataFrame
 Row │ a       b        c
     │ String  Int64?   Array…?
─────┼──────────────────────────
   1 │ x             1  missing
   2 │ y       missing  [1, 2]
```
I would say that this is the most commonly Julia-native pattern needed. If named-tuples were homogenous it would just be `select(df, :col => AsTable)`. Therefore, originally users requested adding an option in `AsTable` allowing to specify the `cols` and `promote` arguments from `push!`.

However, I am hesitant with complicating operation specification syntax even more, therefore for this special case I preferred a separate function.

`flatten=true` is rarely needed in practice because `GroupedDataFrame` in practice gives the user almost an equivalent interface.

File: src/abstractdataframe/nest.jl (hunk 3 above)
Comment (bkamins) [reply src/abstractdataframe/nest.jl:225]: So I propose to just keep the current behavior and add `scalar` kwarg (so keep the implementation I have proposed now).

I know that broadcast mechanics was previously proposed, but it is not correct either. The reason is:
```
julia> length(Base.Broadcast.broadcastable("xyz"))
1
```
so string would be seen as length-1 collection (and this is different from it being seen as a scalar).

Comment (nalimilan) [reply src/abstractdataframe/nest.jl:225]: Yeah, it's a bit tricky. Maybe `length` isn't the right operation. We could check whether `ndims(Base.Broadcast.broadcastable(x)) == 0`, which is the closest equivalent of a scalar in the context of broadcasting. Not sure whether that's worth a breaking change.

Comment (bkamins) [reply src/abstractdataframe/nest.jl:225]: I agree - and for this reason I would just leave the things as they are - just add `scalar` kwarg that is non-breaking.

File: src/abstractdataframe/nest.jl (hunk 1 above)
Comment (nalimilan) [reply src/abstractdataframe/nest.jl:176]: I see. Though I would find it surprising if similar to the example you gave in your previous post `unnest(DataFrame(data=[(a=1:2, b=3:4), (a=5:6, b=7:8)]), :data)` gave a single-row result. At least when a valid Tables.jl table is passed (in particular a `DataFrame`) it's natural that each of its rows gives a row in the result.

Maybe we can treat objects in a different way depending on whether they are tables or rows by default, while providing an argument for users to request explicitly the behavior they want? That would be useful in particular if you happen to have a named tuples of vectors which should not be treated as a table.

Comment (bkamins) [reply src/abstractdataframe/nest.jl:176]: I have thought about this, but then there is a corner case of users mixing tables and non-tables (in one column or across columns).

If we feel `unnest` is useful (I think it is) then - as I have proposed - maybe indeed we need two separate functions: one doing flattening, and other not doing it?

Comment (bkamins) [reply src/abstractdataframe/nest.jl:176]: OK - I have thought about it more and now I have a clear mind about it on an abstract level.

So essentially if you have a nested structure in a cell of a data frame you can ask for the following operations:
1. spread it into multiple rows (this is current `flatten`)
2. spread it into multiple columns if it defines valid column names (we do not have it now; this is the `flatten=false` option in proposed `unnest`)
3. perform both operations 1. and 2. (which makes sense only for tables; this is the `flatten=true` option in proposed `unnest`)

(note that technically some data structures support both 1 and 2: e.g. `NamedTuple` could be spread into multiple rows or multiple columns conceptually)

Given this - maybe indeed we need to add two functions. The verbose names would be `unnest_as_row` and `unnest_as_table` (they are verbose, so probably we should not use them, but they show the intention).

**Comment by bkamins** [comment]
I propose to discuss this PR step by step.
Let us start with `nest`. Here the major question is if we need it. The major reason is that `Ref` is a nesting operator in operator specification syntax.

Start with some data frame:
```
julia> df = DataFrame(a=[1, 1, 2, 2], b=11:14, c=21:24)
4×3 DataFrame
 Row │ a      b      c
     │ Int64  Int64  Int64
─────┼─────────────────────
   1 │     1     11     21
   2 │     1     12     22
   3 │     2     13     23
   4 │     2     14     24
```
First, already doing `groupby` gives us nested `SubDataFrames` - and maybe this is what is enough for most users:
```
julia> groupby(df, :a)
GroupedDataFrame with 2 groups based on key: a
First Group (2 rows): a = 1
 Row │ a      b      c
     │ Int64  Int64  Int64
─────┼─────────────────────
   1 │     1     11     21
   2 │     1     12     22
⋮
Last Group (2 rows): a = 2
 Row │ a      b      c
     │ Int64  Int64  Int64
─────┼─────────────────────
   1 │     2     13     23
   2 │     2     14     24
```
(note that such grouped data frame also supports convenient indexing to get the nested data frames, something which is not easily available if we indeed nest data)

Now, a basic pattern to nest some columns as `NamedTuple` is:
```
julia> combine(groupby(df, :a), AsTable([:b, :c]) => Ref => :bc)
2×2 DataFrame
 Row │ a      bc
     │ Int64  NamedTup…
─────┼─────────────────────────────────────
   1 │     1  (b = [11, 12], c = [21, 22])
   2 │     2  (b = [13, 14], c = [23, 24])
```
The only thing user needs to remember here is that `:b` and `:c` fields of named tuples are views, but this is something that might be preferred in many scenarios.

If one wants to nest data frames one can just write:
```
julia> combine(groupby(df, :a), AsTable([:b, :c]) => Ref∘DataFrame => :bc)
2×2 DataFrame
 Row │ a      bc
     │ Int64  DataFrame
─────┼──────────────────────
   1 │     1  2×2 DataFrame
   2 │     2  2×2 DataFrame
```
Now the columns are copied (as `DataFrame` constructor copies data by default)

Finally note that operation specification syntax also works with row nesting (something we have not discussed but is useful):
```
julia> combine(groupby(df, :a), AsTable([:b, :c]) => ByRow(identity) => :bc)
4×2 DataFrame
 Row │ a      bc
     │ Int64  NamedTup…
─────┼─────────────────────────
   1 │     1  (b = 11, c = 21)
   2 │     1  (b = 12, c = 22)
   3 │     2  (b = 13, c = 23)
   4 │     2  (b = 14, c = 24)
```
In the implementation in the PR I used a bit different pattern:
```
julia> combine(groupby(df, :a), x -> (bc = select(x, [:b, :c]),))
2×2 DataFrame
 Row │ a      bc
     │ Int64  DataFrame
─────┼──────────────────────
   1 │     1  2×2 DataFrame
   2 │     2  2×2 DataFrame
```
but it was only because the implementation should support all cases (and `AsTable` can be problematic with compilation in case of extremely wide tables).

In summary - my question is. Given these considerations do we need to add `nest`? Maybe it is enough to add examples in the manual how nesting can be done?

**Comment by jariji** [comment]
[`combine`](https://dataframes.juliadata.org/stable/lib/functions/#DataFrames.combine) says
> If target_cols is a Symbol or a string then the function is assumed to return a single column. In this case returning a data frame, a matrix, a NamedTuple, or a DataFrameRow raises an error.

e.g. 

```jl
julia> combine(groupby(df, :a), AsTable([:b, :c]) => DataFrame => :bc)
ERROR: ArgumentError: a single value or vector result is required (got DataFrame)
```

Is the reason for this error documented somewhere? i.e. why doesn't this just work?

**Comment by bkamins** [comment]
> i.e. why doesn't this just work?

The reason is safety (in other works: to allow for non-breaking change in the future if we decided it is needed). For example if user writes:
```
AsTable([:b, :c]) => DataFrame
```
it would be ambiguous if user wants the produced data frame to be stored in one cell of a data frame or expanded (as intuitively users might expect it to be expanded just as vectors are expanded).

> The error message could be a good place to document the `Ref/fill` trick.

This is a good point. I will make a PR changing this.

**Comment by nalimilan** [comment]
Yeah `nest` isn't strictly needed. Its main advantage is that it's easier to discover, but it's not clear that nesting is really useful in DataFrames.jl thanks to the existence of `GroupedDataFrame` (that dplyr doesn't have).

That said, if we add `unnest` we should probably have `nest` for consistency. But the behavior with `flatten=false` doesn't really match what I would expect from `unnest`. Maybe the action of splitting a column into several ones would be better named `separate` or `extract` similar to the dplyr functions? In dplyr these only allow splitting strings, but we could make it more general and allow creating columns from any collection, including named tuples? The only peculiarity of named tuples (and dicts) is that appropriate column names can be extracted automatically.

**Comment by bkamins** [comment]
> Maybe the action of splitting a column into several ones would be better named `separate` or `extract` similar to the dplyr functions?

This is what I had in mind. `extract` seems as a good name.

> In dplyr these only allow splitting strings, but we could make it more general and allow creating columns from any collection, including named tuples?

This is something we do not need to add as we already have it. As `AsTable` as target does exactly this.
The only limitation is that `AsTable` assumes a fixed schema for all rows.

What we need is a function designed to handle cases when each row potentially has a different schema.
And maybe also (this is not added, but we could add it) allowing for e.g. `missing` value in a row that would be exapnded to `missing` values.

> The only peculiarity of named tuples (and dicts) is that appropriate column names can be extracted automatically.

Currently dicts would not work as they do not have a defined column order (but we could change this; but then also a change in `push!` et al. should be introduced for consistency)

**Comment by jariji** [comment]
> it's not clear that nesting is really useful in DataFrames.jl thanks to the existence of GroupedDataFrame (that dplyr doesn't have).

I'm missing something, how does GroupedDataFrame substitute for nesting?

**Comment by bkamins** [comment]
> nested dataframes is my workaround for that missing feature.

Indeed nested column labels are not supported and a work-around for them is needed. However, my question is why do you use data frame for this. Normally a `NamedTuple` of scalars would be used here like:
```
julia> df = DataFrame(x=[(a="aa", b="bb"), (a="pp", b="qq")])
2×1 DataFrame
 Row │ x
     │ NamedTup…
─────┼──────────────────────
   1 │ (a = "aa", b = "bb")
   2 │ (a = "pp", b = "qq")
```
That is why I have said that normally I would expect `flatten=false` to be needed.

> I'm missing something, how does `GroupedDataFrame` substitute for df nesting?

What we mean is that you can easily index into a `GroupedDataFrame` to get a portion of the source data frame for certain combination of key column values. Notice that it naturally combines with column nesting of scalars.

The point is that if you nest whole `DataFrame` you fix the row structure. While if you nest a `NamedTuple` of scalars you have nested columns but can `groupby` different columns flexibly later.

**Comment by jariji** [comment]
In my dataframe, each row specifies a regression model and the `:df` column has the data, including the regression residuals, similar to the [broom vignette](https://broom.tidymodels.org/articles/broom_and_dplyr.html). What is your opinion about using this style in DFjl?

**Comment by bkamins** [comment]
You mean that "per group" you want to store different objects:
* source data frame as one column
* estimation results as another column
* residuals etc. as yet another column

Then indeed nesting a data frame (or any other object) makes sense.

**Comment by bkamins** [comment]
@jariji - given my last comment. Now I realized that one cannot create such an object by nesting. I.e. the use case when nesting seems to be indeed needed is when you add different objects sequentially (if you did it in one-shot they would have to have the same number of columns). Is this indeed your case, i.e. you nest only columns needed for estimation of the model but the other columns, that are also nested, are only added later?

File: src/abstractdataframe/nest.jl [hunk 7]
```diff
@@ -0,0 +1,390 @@
+"""
+    nest(gdf::GroupedDataFrame, cols::Pair{<:AbstractString}...)
```
[the remark this answers is not in the document]
Comment (bkamins) [reply src/abstractdataframe/nest.jl:2]: added

File: src/abstractdataframe/nest.jl (hunk 4 above)
Comment (bkamins) [reply src/abstractdataframe/nest.jl:178]: we need because if in some table column is missing we get `missing` initially, which has to get expanded. Now this text is removed.

File: src/abstractdataframe/nest.jl (hunk 1 above)
Comment (bkamins) [reply src/abstractdataframe/nest.jl:179]: I decided to split `unnest` and `expand`

File: src/abstractdataframe/nest.jl [hunk 8]
```diff
@@ -0,0 +1,390 @@
...
+     │ Int64  Int64?   Int64?  
+─────┼─────────────────────────
+   1 │     1        1        5
+   2 │     1        2        6
+   3 │     2  missing  missing 
+   4 │     3  missing        7
+   5 │     3  missing        8
+```
+"""
+function flatten(df::AbstractDataFrame,
```
[the remark this answers is not in the document]
Comment (bkamins) [reply src/abstractdataframe/nest.jl:318]: If you really want it I can do it at the end (but normally I would not do it, as we typically want to squash-merge the PRs, and if we wanted to follow it I would need to rebuild whole git history before the PR is merged, which can be tricky).

I do it this way because abstractdataframe.jl is very large and thus hard to navigate and I think we need to split it into several files (and at some point we need to start doing it).

File: src/abstractdataframe/nest.jl [hunk 9]
```diff
@@ -0,0 +1,390 @@
...
+    return new_df
+end
+
+length_maybe_scalar(v, scalar::Type) = v isa scalar ? -1 : length(v)
+
+function update_lengths!(lengths::Vector{Int}, col::AbstractVector, scalar::Type,
+                         df::AbstractDataFrame, col1, coli)
+    for (i, v) in enumerate(col)
+        lv = length_maybe_scalar(v, scalar)
+        lv == -1 && continue
```
[the remark this answers is not in the document]
Comment (bkamins) [reply src/abstractdataframe/nest.jl:373]: I changed it. (`length_maybe_scalar` was used in one more place, but I changed it also)

File: src/abstractdataframe/nest.jl (hunk 5 above)
Comment (bkamins) [reply src/abstractdataframe/nest.jl:196]: I removed `flatten` so it is simplified. `only` is here to produce `DatraFrameRow`.

**Comment by bkamins** [comment]
OK, I have thought how we should move forward. Although the functions are simple I think we can keep them for user-friendlyness. The functions are:
* `nest` and `unnest` for working with tables
* `expand` and `expand!` for expanding row-like sources (here the `!` version makes sense to have as this is most likely what user will want to do since number of rows does not change).

For now I gave tentative implementations (to show how things internally would work).
@nalimilan To move forward I need https://github.com/JuliaData/DataFrames.jl/pull/3245 to me merged first and then I need to merge it to this PR.

File: src/abstractdataframe/nest.jl [hunk 10]
```diff
@@ -0,0 +1,390 @@
...
+
+Unnest one or more columns `src` into multiple columns. The newly created
+columns are stored at the end of the data frame (and the `src` column is
+dropped).
+
+Each `src` column must contain a `NamedTuple`, a `DataFrameRow`, a
+`Tables.AbstractRow`, or Tables.jl table.
+
+`cols` (default `:setequal`) and `promote` (default `true`) keyword arguments
+have the same meaning as in [`push!`](@ref).
```
[the remark this answers is not in the document]
Comment (bkamins) [reply src/abstractdataframe/nest.jl]: When I started describing it I realized that anything except `cols=:union` and `promote=true` is not really useful. We can always add it later. So I opted for a simpler design for now and functions do not take these arguments.

File: src/abstractdataframe/nest.jl (hunk 6 above)
Comment (nalimilan) [reply src/abstractdataframe/nest.jl:210]: Got it. But why do we need to create a new `DataFrame`? Isn't that inefficient? Can't we just use `append!` instead of `push!` plus `flatten`?

**Comment by nalimilan** [comment]
> > In dplyr these only allow splitting strings, but we could make it more general and allow creating columns from any collection, including named tuples?
> 
> This is something we do not need to add as we already have it. As `AsTable` as target does exactly this. The only limitation is that `AsTable` assumes a fixed schema for all rows.
> 
> What we need is a function designed to handle cases when each row potentially has a different schema. And maybe also (this is not added, but we could add it) allowing for e.g. `missing` value in a row that would be exapnded to `missing` values.

The schema isn't fixed either when splitting string columns into one or more columns: in some cases you might have no occurrence of the separator, in some cases one or more occurrences, giving one, two or more columns. That's why it seems to make sense to be able to support both strings and collections in the same function.

Why call it `expand` rather than `separate` or `extract`? `expand` is something completely different in dplyr.

Otherwise your proposal sounds good to me.

**Comment by jariji** [comment]
I'm reserving comment on the name until I can see it in context with examples.

File: src/abstractdataframe/nest.jl (hunk 6 above)
Comment (bkamins) [reply src/abstractdataframe/nest.jl:210]: right - this is a problem with refactoring :). I have fixed this.

**Comment by bkamins** [comment]
> Why call it `expand` rather than `separate` or `extract`?

I meant `extract` - fixed.

Now regarding:

> The schema isn't fixed either when splitting string columns into one or more columns

This does not supported anyway, as currently there is no way to specify that the string should be split in the syntax.

What we provide is the following:
```
julia> df = DataFrame(x=["a,b", "c,d"])
2×1 DataFrame
 Row │ x
     │ String
─────┼────────
   1 │ a,b
   2 │ c,d

julia> select(df, :x => ByRow(x -> split(x, ',')) => AsTable)
2×2 DataFrame
 Row │ x1         x2
     │ SubStrin…  SubStrin…
─────┼──────────────────────
   1 │ a          b
   2 │ c          d

julia> select(df, :x => ByRow(x -> split(x, ',')) => [:p, :q])
2×2 DataFrame
 Row │ p          q
     │ SubStrin…  SubStrin…
─────┼──────────────────────
   1 │ a          b
   2 │ c          d
```
with the restriction that every string must be split into the same number of groups.

Maybe then - instead of adding `extract` and `extract!` we should change the implementation of `=> AsTable` and `=> [:p, :q]` syntaxes above and allow for varying number of columns to be produced by the expression (in which case we would make the `cols=:union` equivalent instead?).

**Comment by jariji** [comment]
Related PR for string-splitting with fixed number of splitpoints https://github.com/JuliaLang/julia/pull/43557

**Comment by bkamins** [comment]
OK - `flatten` is removed from this PR.
We leave it as WIP with `unnest`, `nest`, `extract` and `extract!`

**Comment by bkamins** [comment]
Self-note. Investigate:
```
:src_column => only => AsTable
```
pattern

**Comment by ohaaga** [comment]
Just learning Julia, so apologies if this is redundant, but I'd love to have nest/unnest (and convenience functions for mapping pipelined functions over columns that contain dataframes) for something like Jennifer Bryan's "row-oriented" workflow, which really helps to keep a project organized (in a real data structure, rather than with ad-hoc naming conventions, etc) when repeating multiple analyses over e.g. different geographic units.

https://github.com/jennybc/row-oriented-workflows

**Comment by seabbs** [comment]
Just bumping this with the same points as the last commenter. I found this via starting from Dataframesmeta and spiralling out. 

The use case I have is here where I ended up doing something pretty clunky 

'''Julia
begin
    # Demonstrate effective censoring windows from combining primary + secondary
    n_samples = 1000000
    secondary_window = 7.0  # Secondary window length

    # DataFramesMeta lacks nest/unnest so we have to get clunky
    censoring_windows_df = @chain scenarios_df begin
        vcat([DataFrame(
                  name = fill(row.name, n_samples),
                  r = fill(row.r, n_samples),
                  color = fill(row.color, n_samples),
                  sample_id = 1:n_samples,
                  primary_time = rand(row.primary_event, n_samples)
              ) for row in eachrow(_)]...)
        @transform :effective_window = secondary_window .- :primary_time
    end
end
'''

From https://censoreddistributions.epiaware.org/dev/getting-started/tutorials/exponentially-tilted-primary-events/

Very open to the idea there are better ways to do this but regardless this nest/unnest pattern is common to a lot of my workflows.

**Comment by bkamins** [comment]
Sure - we can discuss what should be changed. Let us try to pinpoint a minimal set of functionality that would be consistent with the current design and add value.
